src/test-run/index.js

```diff
@@ -0,0 +1,142 @@
+import path from 'path';
+import Promise from 'pinkie';
+import { Session } from 'testcafe-hammerhead';
+import { createDoneCommand, TYPE as COMMAND_TYPE } from './commands';
+import CLIENT_MESSAGES from './client-messages';
+import delay from '../utils/delay';
+import MESSAGE from '../errors/runtime/message';
+import { GeneralError } from '../errors/runtime';
+
+
+export default class TestRun extends Session {
+    constructor (test, browserConnection, screenshotCapturer, opts) {
+        var uploadsRoot = path.dirname(test.fixture.path);
+
+        super(uploadsRoot);
+
+        this.opts               = opts;
+        this.test               = test;
+        this.browserConnection  = browserConnection;
+        this.screenshotCapturer = screenshotCapturer;
+
+        this.started        = false;
+        this.pendingCommand = null;
+        this.pendingRequest = null;
+        this.pendingJsError = null;
+
+        this.errs = [];
+    }
+
+    _getPayloadScript () {
+        // TODO
+    }
+
+    _getIframePayloadScript () {
+        // TODO
+    }
+
+    async _start () {
+        this.started = true;
+
+        try {
+            await this.test.fn(this);
+        }
+        catch (err) {
+            this._fatalError(err);
+        }
+        finally {
+            this._done();
+        }
+    }
+
+    _fatalError (err) {
+        this.errs.push(err);
+    }
+
+    _done () {
+        if (this.pendingRequest)
+            this._resolvePendingRequest(createDoneCommand());
+
+        if (this.pendingJsError)
+            this.errs.push(this.pendingJsError);
+
+        this.emit('done');
+    }
+
+    _resolvePendingCommand () {
+        this.pendingCommand.resolve();
+        this.pendingCommand = null;
+    }
+
+    _rejectPendingCommand (err) {
+        this.pendingCommand.reject(err);
+        this.pendingCommand = null;
+    }
+
+    _resolvePendingRequest (command) {
+        this.pendingRequest.resolve(command);
+        this.pendingRequest = null;
+    }
+
+    getAuthCredentials () {
+        // TODO
+    }
+
+    handleFileDownload () {
+        // TODO
+    }
+
+    handlePageError () {
+        // TODO
+    }
+
+    executeCommand (command) {
+        if (this.pendingCommand)
+            return Promise.reject(new GeneralError(MESSAGE.attemptToRunSeveralCommands));
```

Comment:
It shouldn't be GeneralError and it shouldn't be promise. It's an internal assertion, this should break TestCafe and we can write text inline here. This is just guard for us to avoid regressions. See https://en.wikipedia.org/wiki/Assertion_(software_development)